File: src/hpss.cpp
#include "hpss.h"
#include "fft_utils.h"
#include <iostream>
#include <algorithm>
#include <limits>

namespace HPSS
{

using namespace fft_utils;
using Vec2D = std::vector<std::vector<float>>;

/** Performs "horizontal" median filtering to obtain harmonic signal */
Vec2D median_filter_harm(const std::vector<fftw_complex_vec>& S, int kernel_size)
{
    Vec2D H (S.size(), std::vector<float> (S[0].size(), 0.0f));

    const int pad = (kernel_size - 1) / 2;
    std::vector<float> med_vec (S.size() + 2 * pad, 0.0f);
    std::vector<float> kernel_vec (kernel_size, 0.0f);

    for(int h = 0; h < (int) S[0].size(); ++h)
    {
        for(int i = 0; i < (int) S.size(); ++i)
            med_vec[i + pad] = std::abs(S[i][h]);

        for(int i = 0; i < (int) S.size(); ++i)
        {
            std::copy(med_vec.begin() + i, med_vec.begin() + i + kernel_size, kernel_vec.begin());
            std::nth_element(kernel_vec.begin(), kernel_vec.begin() + pad, kernel_vec.end());
            H[i][h] = kernel_vec[pad];
        }
    }

    return H;
}

/** Performs "vertical" median filtering to obtain percussive signal */
Vec2D median_filter_perc(const std::vector<fftw_complex_vec>& S, int kernel_size)
{
    Vec2D P (S.size(), std::vector<float> (S[0].size(), 0.0f));

    const int pad = (kernel_size - 1) / 2;
    std::vector<float> med_vec (S[0].size() + 2 * pad, 0.0f);
    std::vector<float> kernel_vec (kernel_size, 0.0f);

    for(int i = 0; i < (int) S.size(); ++i)
    {
        for(int h = 0; h < (int) S[i].size(); ++h)
            med_vec[h + pad] = std::abs(S[i][h]);

        for(int h = 0; h < (int) S[i].size(); ++h)
        {
            std::copy(med_vec.begin() + i, med_vec.begin() + i + kernel_size, kernel_vec.begin());
            std::nth_element(kernel_vec.begin(), kernel_vec.begin() + pad, kernel_vec.end());
            P[i][h] = kernel_vec[pad];
        }
    }

    return P;
}

/** Computes a time-frequency spectrogram with no window (for now) */
inline std::vector<fftw_complex_vec> spectrogram(const std::vector<float>& x, int fft_size, int hop_size, int zero_pad = 1)
{
    const auto n_fft = fft_size * zero_pad;

    std::vector<fftw_complex_vec> S;
    fft_utils::ForwardFFT fft { n_fft };
    for(int i = 0; i + fft_size < (int) x.size(); i += hop_size)
    {
        std::copy(x.begin() + i, x.begin() + i + fft_size, fft.x_in.data());
        fft.perform();
        S.push_back(fft.Y_out);
    }

    return S;
}

/** Reconstructs the harmonic and percussive signals from their spectrograms with a Hann window */
std::pair<std::vector<float>, std::vector<float>> spec_reconstruct(std::vector<fftw_complex_vec>& H_hat,
                                                                   std::vector<fftw_complex_vec>& P_hat,
                                                                   int n_samples,
                                                                   const int fft_size,
                                                                   const int hop_size)
{
    std::vector<float> h_sig (n_samples, 0.0f);
    std::vector<float> p_sig (n_samples, 0.0f);
    const auto win = hann(fft_size, float(fft_size / hop_size) / 2.0f);

    const int n_fft = (int) H_hat[0].size();
    fft_utils::InverseFFT ifft { n_fft };
    for(int i = 0; i < H_hat.size(); ++i)
    {
        int start_idx = i * hop_size;
        int samples = std::min(fft_size, n_samples - start_idx);

        { // do H
            std::copy(H_hat[i].begin(), H_hat[i].end(), ifft.X_in.data());
            ifft.perform();
            fft_utils::applyWindow(ifft.y_out, win, ifft.y_out);

            for(int n = 0; n < samples; ++n)
                h_sig[n + start_idx] += ifft.y_out[n];
        }

        { // do P
            std::copy(P_hat[i].begin(), P_hat[i].end(), ifft.X_in.data());
            ifft.perform();
            fft_utils::applyWindow(ifft.y_out, win, ifft.y_out);

            for(int n = 0; n < samples; ++n)
                p_sig[n + start_idx] += ifft.y_out[n];
        }
    }

    return std::make_pair(h_sig, p_sig);
}

static void debug_print(const std::string& str, bool debug)
{
    if(debug)
        std::cout << str << std::endl;
}

std::pair<std::vector<float>, std::vector<float>> hpss(std::vector<float> x, const HPSS_PARAMS& params)
{
    const auto fft_size = next_pow2(int(params.window_length_ms * 0.001 * params.sample_rate));
    const auto hop_size = next_pow2(int(params.hop_length_ms * 0.001 * params.sample_rate));

    debug_print("Computing HPSS...", params.debug);
    debug_print("\tComputing STFTs...", params.debug);
    auto S = spectrogram(x, fft_size, hop_size, params.zero_pad);

    debug_print("\tSeparating percussive signal...", params.debug);
    auto P = median_filter_perc(S, params.perc_kernel);

    debug_print("\tSeparating harmonic signal...", params.debug);
    auto H = median_filter_harm(S, params.harm_kernel);

    debug_print("\tApplying filter masks...", params.debug);
    std::vector<fftw_complex_vec> H_hat (S.size(), fftw_complex_vec(S[0].size()));
    std::vector<fftw_complex_vec> P_hat (S.size(), fftw_complex_vec(S[0].size()));
    for(int i = 0; i < (int) S.size(); ++i)
    {
        for(int h = 0; h < (int) S[0].size(); ++h)
        {
            float H_p = std::pow(H[i][h], params.mask_exp);
            float P_p = std::pow(P[i][h], params.mask_exp);
            float denom = H_p + P_p + std::numeric_limits<float>::epsilon();

            H_hat[i][h] = S[i][h] * (H_p / denom);
            P_hat[i][h] = S[i][h] * (P_p / denom);
        }
    }

    debug_print("\tComputing time-domain signal...", params.debug);
    return spec_reconstruct(H_hat, P_hat, (int) x.size(), fft_size, hop_size);
}

} // namespace HPSS

File: python/hpss.py
import numpy as np
from scipy.signal import windows

eps = np.finfo(np.float32).eps

def median_filter(x, kernel_size):
    pad = (kernel_size - 1) // 2
    x_pad = np.concatenate((np.zeros(pad), x, np.zeros(pad)))
    y = np.zeros_like(x)
    for i in range(len(y)):
        median_arg = np.argsort(np.abs(x_pad[i : i + kernel_size]))[pad]
        y[i] = x_pad[i + median_arg]

    return y

def spectrogram(x, fft_size, hop_size, zero_pad=1):
    S = None
    for i in range(0, len(x), hop_size):
        x_win = x[i : i + fft_size]
        x_pad = np.zeros(fft_size * zero_pad)
        x_pad[:len(x_win)] = x_win

        if S is None:
            S = np.array([np.fft.fft(x_pad)])
        else:
            S = np.append(S, np.array([np.fft.fft(x_pad)]), axis=0)

    return S

def hpss(x, perc_kernel=17, harm_kernel=17, mask_power=2, fft_size=4096, hop_size=1024, zero_pad=2):
    ''' Simple harmonic/percussive source separation based on median filter method '''

    print('Computing HPSS...')
    print('\tComputing STFTs...')
    S = spectrogram(x, fft_size, hop_size, zero_pad)

    # percussive signal
    print('\tSeparating percussive signal...')
    P = np.copy(S)
    for i in range(S.shape[0]):
        P[i, :] = median_filter(np.abs(S[i, :]), kernel_size=perc_kernel)

    # harmonic signal
    print('\tSeparating harmonic signal...')
    H = np.copy(S)
    for h in range(S.shape[1]):
        H[:, h] = median_filter(np.abs(S[:, h]), kernel_size=harm_kernel)

    # create filter masks
    print('\tCreating filter masks...')
    M_H = np.copy(S)
    M_P = np.copy(S)
    for i in range(S.shape[0]):
        for h in range(S.shape[1]):
            H_p = H[i,h]**mask_power
            P_p = P[i,h]**mask_power
            denom = H_p + P_p + eps

            M_H[i, h] = H_p / denom
            M_P[i, h] = P_p / denom

    H_hat = np.multiply(S, M_H)
    P_hat = np.multiply(S, M_P)

    print('\tComputing time-domain signal...')
    h_sig = np.zeros_like(x)
    p_sig = np.zeros_like(x)
    for i in range(S.shape[0]):
        start_idx = int(i * hop_size)
        n_samples = min(fft_size, len(x) - start_idx)
        win = windows.hann(fft_size)[:n_samples] / ((fft_size // hop_size) // 2)
        h_sig[start_idx : start_idx + fft_size] += win * np.real(np.fft.ifft(H_hat[i,:])[:n_samples])
        p_sig[start_idx : start_idx + fft_size] += win * np.real(np.fft.ifft(P_hat[i,:])[:n_samples])

    return h_sig, p_sig
